File: BTree/BTree.py
class BTreeNode:
    def __init__(self, leaf=False):
        self.leaf = leaf
        self.keys = []
        self.children = []

class BTree:
    def __init__(self, t):
        self.root = BTreeNode(True)
        self.t = t  # 최소 차수

    def search(self, node, key):
        i = 0
        while i < len(node.keys) and key > node.keys[i]:
            i += 1
        if i < len(node.keys) and key == node.keys[i]:
            return (node, i)
        elif node.leaf:
            return None
        else:
            return self.search(node.children[i], key)

    def insert(self, key):
        root = self.root
        if len(root.keys) == 2 * self.t - 1:
            s = BTreeNode()
            self.root = s
            s.children.append(root)
            self.split_child(s, 0)
            self.insert_non_full(s, key)
        else:
            self.insert_non_full(root, key)

    def insert_non_full(self, node, key):
        i = len(node.keys) - 1
        if node.leaf:
            node.keys.append(0)
            while i >= 0 and key < node.keys[i]:
                node.keys[i + 1] = node.keys[i]
                i -= 1
            node.keys[i + 1] = key
        else:
            while i >= 0 and key < node.keys[i]:
                i -= 1
            i += 1
            if len(node.children[i].keys) == 2 * self.t - 1:
                self.split_child(node, i)
                if key > node.keys[i]:
                    i += 1
            self.insert_non_full(node.children[i], key)

    def split_child(self, parent, i):
        t = self.t
        node = parent.children[i]
        new_node = BTreeNode(node.leaf)
        parent.children.insert(i + 1, new_node)
        parent.keys.insert(i, node.keys[t - 1])
        new_node.keys = node.keys[t:(2 * t - 1)]
        node.keys = node.keys[0:(t - 1)]
        if not node.leaf:
            new_node.children = node.children[t:(2 * t)]
            node.children = node.children[0:(t - 1)]

    def inorder_traversal(self, node):
        i = 0
        for i in range(len(node.keys)):
            if not node.leaf:
                self.inorder_traversal(node.children[i])
            print(node.keys[i], end=" ")
        if not node.leaf:
            self.inorder_traversal(node.children[i + 1])

# Usage
b_tree = BTree(3)
values = [10, 20, 5, 6, 12, 30, 7, 17]
for v in values:
    b_tree.insert(v)

b_tree.inorder_traversal(b_tree.root)  # Output: 5 6 7 10 12 17 20 30

File: BTree/BTree.cs
using System;
using System.Collections.Generic;

class BTreeNode
{
    public List<int> Keys;
    public List<BTreeNode> Children;
    public bool Leaf;

    public BTreeNode(bool leaf)
    {
        Leaf = leaf;
        Keys = new List<int>();
        Children = new List<BTreeNode>();
    }
}

class BTree
{
    private int T;
    private BTreeNode root;

    public BTree(int t)
    {
        T = t;
        root = new BTreeNode(true);
    }

    public void Insert(int key)
    {
        BTreeNode r = root;
        if (r.Keys.Count == 2 * T - 1)
        {
            BTreeNode s = new BTreeNode(false);
            root = s;
            s.Children.Add(r);
            SplitChild(s, 0);
            InsertNonFull(s, key);
        }
        else
        {
            InsertNonFull(r, key);
        }
    }

    private void InsertNonFull(BTreeNode node, int key)
    {
        int i = node.Keys.Count - 1;
        if (node.Leaf)
        {
            node.Keys.Add(0);
            while (i >= 0 && key < node.Keys[i])
            {
                node.Keys[i + 1] = node.Keys[i];
                i--;
            }
            node.Keys[i + 1] = key;
        }
        else
        {
            while (i >= 0 && key < node.Keys[i])
                i--;
            i++;
            if (node.Children[i].Keys.Count == 2 * T - 1)
            {
                SplitChild(node, i);
                if (key > node.Keys[i])
                    i++;
            }
            InsertNonFull(node.Children[i], key);
        }
    }

    private void SplitChild(BTreeNode parent, int i)
    {
        BTreeNode y = parent.Children[i];
        BTreeNode z = new BTreeNode(y.Leaf);
        parent.Children.Insert(i + 1, z);
        parent.Keys.Insert(i, y.Keys[T - 1]);

        z.Keys.AddRange(y.Keys.GetRange(T, T - 1));
        y.Keys.RemoveRange(T - 1, T);

        if (!y.Leaf)
        {
            z.Children.AddRange(y.Children.GetRange(T, T));
            y.Children.RemoveRange(T, T);
        }
    }

    public void InorderTraversal(BTreeNode node)
    {
        for (int i = 0; i < node.Keys.Count; i++)
        {
            if (!node.Leaf)
                InorderTraversal(node.Children[i]);
            Console.Write(node.Keys[i] + " ");
        }
        if (!node.Leaf)
            InorderTraversal(node.Children[node.Keys.Count]);
    }

    public BTreeNode GetRoot() => root;
}

class Program
{
    static void Main()
    {
        BTree bTree = new BTree(3);
        int[] values = { 10, 20, 5, 6, 12, 30, 7, 17 };
        foreach (var value in values)
            bTree.Insert(value);

        bTree.InorderTraversal(bTree.GetRoot());  // Output: 5 6 7 10 12 17 20 30
    }
}